### ISSUE TITLE
Better Backfill UX

### ISSUE THREAD

**Issue by flarco** [slingdata-io/sling-cli#41]
Currently, when backfilling very large tables, there is no way to incrementally backfill without manual intervention. 

Some ideas on improving the ease of backfilling:
- When mode is `incremental`, have some option, such as`flush_interval`, defaulting to 10M records, so that records are written in the final table incrementally instead of copying the full table.
  - This will enable some resilience, in case some error occurs mid-transfer of very large tables. We would not have to start from scratch.
- Have an input for a range of backfill for the `update_key`? This will allow specific backfill periods.
- Backfills are usually ad-hoc operations. Perhaps have `backfill` mode, which is really `incremental` mode, with a specific range for the `update_key`.
  - `primary_key` and `update_key` would be required. 
  - `range` could be `all`, which is really `incremental`
  - have ability to specify particular columns that we want to backfill.

**Comment by nixent** [comment]
Having separate `backfill` mode would be better option. Normally `incremental` mode is either runs full load on first run and then continues with incremental load on subsequent runs or runs incremental load from first run with backfill done separately.

**Comment by GJMcClintock** [comment]
@flarco love the tool - it was so simple to spin up and test. Was going to reach out about this because we are staying down a table that will probably take 16-20 hours to backfill and the current prop is all or nothing.

- I think the `flush_interval` option is a great idea. Meltano does similar
- I was thinking of a general "max load" size rather than a flush could also work. Then if it doesn't load in one run, that's fine (similar to the airflow backfil paradigm). This was actually the expected behavior I had by using the `limit` source option. Unless you have a source where the number of records at a single timestamp is greater than the limit the `>= max(updated_key) limit {value]` should always work with some redundancy if hte limit hits at a point where there's a tie.

**Comment by GJMcClintock** [comment]
Another option as I sit here thinking about it could be an option bound for the incremental condition -

eg: if the update_key is an int, you could say 1,000,000 making it `where id >= max(id) and id <= max(id) + 1000000)

or if it's a date, `>= max(date) and <= max(date) + 3` to grab 3 day chunks.

for the dates, this could get stuck in a loop if you don't have records for X days. This could be handled either with logging to kick out the date if it's the same as last time, OR to try again dropping the upper bound if no records are returned, ie:

`>= max(date) and <= max(date) + 3` returns nothing, run it against wtih `>= max(date)` to see if there's ANYTHING newer.

**Comment by nixent** [comment]
In our setup, we wanted backfills to be deterministic, meaning that backfill should run on requested range and on that range only. If it doesn't return any records, that is fine, just request another range. This way data are partitioned by grain of `update_key` and is very useful for cases when we find errors in source system, fix it and backfill just a one or few partitions. Data corrections in the source can be deletes, so `incremental` on `update_key` will not capture it. 
So far we use `full-refresh` mode with custom SQL statement in the `stream` to implement range logic. It would be nice to have separate backfill mode but it is more of UX as the title says.

**Comment by GJMcClintock** [comment]
@nixent - interesting appraoch and makes sense. I'm assuming you have a step prior to the run to generate your needed range and then pass that in? that's easy enough to set up albeit not as turnkey.

supporting partitioning with the incremental range would also be a nice feature - right now each commit in snowflake would make a micro partition in the temp table and there isn't a setting for setting an order by on the final commit to force it to naturally partition along that column.

**Comment by nixent** [comment]
Yes, the range script is generated by orchestrator, Dagster. 
Dagster allows selecting partitions for backfill. You can select one partition, range of partitions or all of them. Depending on backfill policy, multiple backfills can run in parallel with respective concurrency limits or backfills can run sequentially, one by one. Backfills can also run on schedule to use resources in off peak time.
I think it is important to separate responsibilities of different components of data platform, so Sling is doing EL and orchestrator is doing orchestration. Sling suits very well for embedding into orchestrator and it might be tempting to implement some of orchestration in Sling, however I would not recommend that.

**Comment by flarco** [comment]
Thanks @GJMcClintock @nixent.

Here is what I have in mind to get something out quick (see PR https://github.com/slingdata-io/sling-cli/pull/95), notably adding `--limit` and `--range` flags in the CLI, mapped to `source.options.limit` and `source.options.range` respectively, as well as adding `backfill` mode:

**Backfill range 2021-01-01 to 2021-02-01**
`sling run --src-conn my_source --src-stream schema.mytable --tgt-conn my_target --tgt-object schema.new_table --primary-key id --update-key updated_at --mode backfill --range 2021-01-01,2021-02-01`

**Backfill range 1 to 10000**
`sling run --src-conn my_source --src-stream schema.mytable --tgt-conn my_target --tgt-object schema.new_table --primary-key id --update-key updated_at --mode backfill --range 1,10000`

**Incremental limited**
`sling run --src-conn my_source --src-stream schema.mytable --tgt-conn my_target --tgt-object schema.new_table --primary-key id --update-key updated_at --mode incremental --limit 10000000`

For `backfill` mode, we must specify a valid range value (with `--range` or `source.options.range`) separated by one comma `{start},{end}`.  Basically, the where clause will be constructed as `{update_key} >= {start_value} and {update_key} <= {end_value}` (see [here](https://github.com/slingdata-io/sling-cli/pull/95/files#diff-52b3cf4f4953fc17c2c73c4532691012530c85a1e9961ccf4d945ef9b78e1c41R87-R103) for logic)

For `--limit`, this will be useful for limiting incrementally, as @GJMcClintock mentioned. Actually, you could simply run that on loop until all data is transferred. Notable modification is that now SELECT query will be ordered with `order by {update_key} asc`.

I think this should do the trick. `flush_interval` will have to wait, as that requires heavy refactoring. I think `--limit` will mostly supplement that function, just have to run the command again, where `flush_interval` would do it by itself in-process.

Thoughts? Suggestions? I'll probably release this weekend.

**Comment by GJMcClintock** [comment]
I think this is great - thanks for being so responsive!

**Comment by flarco** [comment]
v1.0.66 released!
